content/curried-varargs.md

```diff
@@ -0,0 +1,271 @@
+2. Unnecessary temporary `Seq` to hold repeated parameters.
+3. Unnecessary runtime type check for each argument.
+
+The similar issues can be found in many other usage of repeated parameters. For example, Scala string interpolation is inefficient due to its internal vararg function call, unless implementing it in a macro; Scala collection initializers (e.g. `List(1, 2, 3)`) create unnecessary temporary `Seq` before creating the desired collection.
+
+## Design
+
+This proposal introduces a new type `Curried` defined as following:
+
+``` scala
+trait Curried extends Any
+```
+
+When a function call `f(p1, p2, p3, ... pn)` is being type checked, the compiler will firstly look for `apply` method on `f`. If an applicable `apply` method is not found and `f` is a subtype of `Curried`, the compiler will convert the function call to curried form `f.applyBegin.applyNext(p1).applyNext(p2).applyNext(p3) ... .applyNext(pn).applyEnd`, and continue type checking the translated call.
+
+### Expanding sequence argument
+
+Optionally, some arguments to a `Curried` call may be a sequence argument marked as  `_*`. Those are arguments should be translated to `applyNextSeq` calls instead of `applyNext`. For example, `f(p1, s1: _*, p2)` will be translated to the following code.
+
+``` scala
+f.applyBegin
+  .applyNext(p1)
+  .applyNextSeq(s1)
+  .applyNext(p2)
+.applyEnd
+```
+
+Unlike traditional repeated parameters, which restrict the sequence argument at the last position, sequence arguments in a curried call are allowed at any position.
+
+### Builder type shifting
+
+The type of partially applied function might be changed during applying each argument. Given the following type signature:
+
+``` scala
+class ListBuilder[A] {
+  def applyNext[B >: A](b: B): ListBuilder[B] = ???
+  def applyNextSeq[B >: A](seqB: Seq[B]): ListBuilder[B] = ???
+  def applyEnd: List[A] = ???
+}
+object List extends Curried {
+  def applyBegin[A]: ListBuilder[A] = ???
+}
+```
+
+`List(42, "a")` should be translated to `List.applyBegin.applyNext(42).applyNext("a").applyEnd`. Then, the typer will infer type parameters as `List.applyBegin[Nothing].applyNext[Int](42).applyNext[Any]("a").applyEnd`, therefore the final return type of `applyEnd` will be `List[Any]`.
+
+### Explicit type parameters
+
+When a `Curried` is invoked with some type arguments, those type arguments will be moved to the `applyBegin` method. Therefore, `List[Int](1 to 3: _*)` should be translated to `List.applyBegin[Int].applyNextSeq(1 to 3).applyEnd`.
+
+### Implicit parameters
+
+A more common form of curried function call would be like `f(a)(b)(c)`. We prefer the explicit named method calls to `applyNext` instead of the common form, in order to support implicit parameters in `applyNext`. Therefore, each explicit parameter might come with an implicit parameter list, resolving the infamous [multiple type parameter lists](https://github.com/scala/bug/issues/4719) issue.
```

Comment:
I don't understand how the multiple type parameter lists problem comes into play here. Perhaps multiple `using` param lists (?), but that is already fixed in Scala 3.

The multiple type param list problem itself would be fixed by [Clause interweaving](https://contributors.scala-lang.org/t/clause-interweaving-allowing-def-f-t-x-t-u-y-u/5525).

Reply:
> Perhaps multiple using param lists (?), but that is already fixed in Scala 3.

 Right. This document was written before Scala 3.